xdsl/dialects/experimental/dmp.py

```diff
@@ -0,0 +1,436 @@
+    """
+    This represents shape information that is attached to halo operations.
+
+    On the terminology used:
+
+    In each dimension, we are given four points. We abbreviate them in
+    annotations to an, bn, cn, dn, with n being the dimension. In 2d, these
+    create the following pattern, higher dimensional examples can
+    be derived from this:
+
+    a0 b0          c0 d0
+    +--+-----------+--+ a1
+    |  |           |  |
+    +--+-----------+--+ b1
+    |  |           |  |
+    |  |           |  |
+    |  |           |  |
+    |  |           |  |
+    +--+-----------+--+ c1
+    |  |           |  |
+    +--+-----------+--+ d1
+
+    We can now name these points:
+
+         - a: buffer_start
+         - b: core_start
+         - c: core_end
+         - d: buffer_end
+
+    This class provides easy getters for these four.
+
+    We can also define some common sizes on this object:
+
+        - buff_size(n) = dn - an
+        - core_size(n) = cn - bn
+        - halo_size(n, start) = bn - an
+        - halo_size(n, end  ) = dn - cn
+    """
+
+    name = "dmp.halo_bounds"
+
+    buff_lb_: ParameterDef[builtin.DenseArrayBase]
+    buff_ub_: ParameterDef[builtin.DenseArrayBase]
+    core_lb_: ParameterDef[builtin.DenseArrayBase]
+    core_ub_: ParameterDef[builtin.DenseArrayBase]
+
+    # helpers for named dimensions:
+    DIM_X: ClassVar[int] = 0
+    DIM_Y: ClassVar[int] = 1
+    DIM_Z: ClassVar[int] = 2
+
+    @property
+    def buff_lb(self) -> tuple[int, ...]:
+        data = self.buff_lb_.as_tuple()
+        assert isa(data, tuple[int, ...])
+        return data
+
+    @property
+    def buff_ub(self) -> tuple[int, ...]:
+        data = self.buff_ub_.as_tuple()
+        assert isa(data, tuple[int, ...])
+        return data
+
+    @property
+    def core_lb(self) -> tuple[int, ...]:
+        data = self.core_lb_.as_tuple()
+        assert isa(data, tuple[int, ...])
+        return data
+
+    @property
+    def core_ub(self) -> tuple[int, ...]:
+        data = self.core_ub_.as_tuple()
+        assert isa(data, tuple[int, ...])
+        return data
+
+    @property
+    def dims(self) -> int:
+        return len(self.core_ub)
+
+    @staticmethod
+    def from_index_attrs(
+        buff_lb: stencil.IndexAttr,
+        core_lb: stencil.IndexAttr,
+        core_ub: stencil.IndexAttr,
+        buff_ub: stencil.IndexAttr,
+    ):
+        # translate everything to "memref" coordinates
+        buff_lb_tuple = (buff_lb - buff_lb).as_tuple()
+        buff_ub_tuple = (buff_ub - buff_lb).as_tuple()
+        core_lb_tuple = (core_lb - buff_lb).as_tuple()
+        core_ub_tuple = (core_ub - buff_lb).as_tuple()
+
+        typ = builtin.i64
+        return HaloShapeInformation(
+            [
+                builtin.DenseArrayBase.from_list(typ, data)
+                for data in (buff_lb_tuple, buff_ub_tuple, core_lb_tuple, core_ub_tuple)
+            ]
+        )
+
+    def buffer_start(self, dim: int) -> int:
```

Comment:
Lots of duplicated code..

Reply:
I can't see a way to factorize this that makes it more readable though.

This is an already super complex piece of code, and I don't want to add more indirection.

Reply:
ok, maybe we can  imporove later